### PR TITLE
feat: construction endpoints

### PR DESCRIPTION
**Pull request by AlexBHarley**
This PR represents my investigation into the construction API, the goal was to get the `rosetta-cli check:construction` command working.

I've added some files to a `tests/` folder, you should be able to run Celo Rosetta and then `rosetta-cli check:construction --configuration-file ./tests/config.json` to verify it works as expected.

### CODE REVIEW

File: service/rpc/servicer.go [hunk 1]
```diff
@@ -419,50 +424,228 @@ func (s *Servicer) BlockTransaction(ctx context.Context, request *types.BlockTra
...
 func (s *Servicer) ConstructionPreprocess(
-	context.Context,
-	*types.ConstructionPreprocessRequest,
+	ctx context.Context,
+	req *types.ConstructionPreprocessRequest,
 ) (*types.ConstructionPreprocessResponse, *types.Error) {
-	return nil, LogErrUnimplemented("ConstructionPreprocess")
+	options := make(map[string]interface{})
+	options["From"] = req.Operations[0].Account.Address
+	options["To"] = req.Operations[1].Account.Address
```
Comment (eelanagaraj) [inline service/rpc/servicer.go:642]: (? checking my understanding): I'm not totally sure if we need to do anything for `preprocess`...do we actually need to make a call to `/construction/metadata` at all? according to the docs we don't need to return any options if we don't need to make a call to `/construction/metadata` and I don't understand why we would need to get From/To addresses online?

File: tests/config.json [hunk 2]
```diff
@@ -0,0 +1,77 @@
+{
+  "network": {
```
Comment (eelanagaraj) [inline tests/config.json:2]: FYI there's also a `cli-config.json` file in the repo which is an older version of this (missing construction API stuff); maybe replace that file with this updated one or just delete that one as well?

**Review by eelanagaraj — COMMENTED** [review]
Wrote mostly a bunch of questions to make sure I'm understanding stuff + some smaller stuff here and there!

File: .golangci.yml [hunk 3]
```diff
@@ -19,7 +19,7 @@ linters:
     - varcheck
 
     - goconst
-    - goimports
+    # - goimports
```
[the remark this answers is not in the document]
Comment (AlexBHarley) [reply .golangci.yml:22]: I couldn't get the linter to pass with it

File: service/rpc/servicer.go (hunk 1 above)
Comment (AlexBHarley) [reply service/rpc/servicer.go:642]: Why would we not make a fetch to `/metadata`? These values are used in the `ConstructionMetadata` endpoint

File: service/rpc/servicer.go [hunk 4]
```diff
@@ -419,50 +424,228 @@ func (s *Servicer) BlockTransaction(ctx context.Context, request *types.BlockTra
...
+			Type: "transfer",
+			OperationIdentifier: &types.OperationIdentifier{
+				Index: 0,
+			},
+			Account: &types.AccountIdentifier{
+				Address: tx.From.Hex(),
+			},
+			Amount: &types.Amount{
+				Value:    new(big.Int).Neg(tx.Value).String(),
+				Currency: CeloGold,
```
[the remark this answers is not in the document]
Comment (AlexBHarley) [reply service/rpc/servicer.go:548]: Definitely. This was a proof of concept to get something working. I always find it easier to work with a project when there are _some_ passing tests. Now we have a passing CELO transfer, we can add failing cUSD ones and it'll be clear what's going wrong.

File: service/rpc/servicer.go [hunk 5]
```diff
@@ -419,50 +424,228 @@ func (s *Servicer) BlockTransaction(ctx context.Context, request *types.BlockTra
...
+	if err := json.Unmarshal([]byte(req.UnsignedTransaction), &metadata); err != nil {
+		return nil, ErrInternal
+	}
+
+	tx := airgap.Transaction{
+		TxMetadata: &metadata,
+		Signature: req.Signatures[0].Bytes,
+	}
+
+	signedTx, err := tx.AsGethTransaction()
```
[the remark this answers is not in the document]
Comment (AlexBHarley) [reply service/rpc/servicer.go:460]: Yeah it has to do basically with how we're shuffling data around. One is `rlpEncode`'d when serialised, and the other not. I could only get it to work the way it is.

File: service/rpc/servicer.go [hunk 6]
```diff
@@ -419,50 +424,228 @@ func (s *Servicer) BlockTransaction(ctx context.Context, request *types.BlockTra
...
-	*types.ConstructionDeriveRequest,
-) (*types.ConstructionDeriveResponse, *types.Error) {
-	return nil, LogErrUnimplemented("ConstructionDerive")
+	if err := json.Unmarshal([]byte(req.UnsignedTransaction), &metadata); err != nil {
+		return nil, ErrInternal
+	}
+
+	tx := airgap.Transaction{
+		TxMetadata: &metadata,
+		Signature: req.Signatures[0].Bytes,
```
[the remark this answers is not in the document]
Comment (AlexBHarley) [reply service/rpc/servicer.go:457]: Agreed. Again proof of concept

File: service/rpc/servicer.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (eelanagaraj) [reply service/rpc/servicer.go:551]: Ok great, thanks for clarifying!

File: service/rpc/servicer.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (eelanagaraj) [reply service/rpc/servicer.go:639]: sweet! yeah this isn't blocking merging or anything, more wondering wrt cUSD design. thanks for the clarification!

File: service/rpc/servicer.go [hunk 7]
```diff
@@ -419,50 +424,228 @@ func (s *Servicer) BlockTransaction(ctx context.Context, request *types.BlockTra
...
 ) (*types.ConstructionParseResponse, *types.Error) {
-	return nil, LogErrUnimplemented("ConstructionParse")
+	var tx airgap.Transaction
+	if !req.Signed {
+		err := json.Unmarshal([]byte(req.Transaction), &tx)
+		if err != nil {
+			return nil, ErrInternal
+		}
+	} else {
+		t := new(ethTypes.Transaction)
```
[the remark this answers is not in the document]
Comment (AlexBHarley) [reply service/rpc/servicer.go:595]: Just re ran these again and yeah, doesn't seem to work. That part of the postman workflow sucks, the tooling to generate a signature quickly is all our own scripts right now.

File: service/rpc/servicer.go (hunk 1 above)
Comment (eelanagaraj) [reply service/rpc/servicer.go:733]: also not blocking merge for this proof of concept, but i think my understanding of this endpoint within the workflow is shaky. My question was more geared towards how the docs describe this endpoint's function and i think i confused myself a bit. Again not blocking, but we may also need to add handling for `Operations.amount.value` but that can come later (and prob goes nicely with the currency changes described above).

**Review by eelanagaraj — APPROVED** [review]

File: service/rpc/servicer.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (eelanagaraj) [reply service/rpc/servicer.go:595]: ah ok, hmm yeah we can keep that in mind and see if we find a better way going forward

**Pull request merged**